File: leetcode/two-pointers/792.匹配子序列的单词数.ts
/*
 * @lc app=leetcode.cn id=792 lang=typescript
 *
 * [792] 匹配子序列的单词数
 *
 * https://leetcode-cn.com/problems/number-of-matching-subsequences/description/
 *
 * algorithms
 * Medium (31.84%)
 * Likes:    91
 * Dislikes: 0
 * Total Accepted:    4.6K
 * Total Submissions: 10.9K
 * Testcase Example:  '"abcde"\n["a","bb","acd","ace"]'
 *
 * 给定字符串 s 和单词字典 words, 求 words[i] 中是 S 的子序列的单词个数。
 *
 *
 * 示例:
 * 输入:
 * s = "abcde"
 * words = ["a", "bb", "acd", "ace"]
 * 输出: 3
 * 解释: 有三个是 S 的子序列的单词: "a", "acd", "ace"。
 *
 *
 * 注意:
 *
 *
 * 所有在words和 S 里的单词都只由小写字母组成。
 * s 的长度在 [1, 50000]。
 * words 的长度在 [1, 5000]。
 * words[i]的长度在[1, 50]。
 *
 *
 */

// @lc code=start
// two pointers
var numMatchingSubseq = function (s: string, words: string[]): number {
  let res = 0;
  for (let i = 0; i < words.length; i++) {
    let j = 0;
    let k = 0;
    while (j < words[i].length && k < s.length) {
      if (words[i][j] === s[k]) {
        j++;
      }
      k++;
    }
    if (j === words[i].length) {
      res++;
    }
  }
  return res;
};

// two pointers
var numMatchingSubseq = function (s: string, words: string[]): number {
  const bucket: [number, number][][] = Array.from(new Array(26), () => []);
  for (let i = 0; i < words.length; i++) {
    bucket[words[i].charCodeAt(0) - 97].push([i, 0]);
  }
  let res = 0;
  for (const c of s) {
    const len = bucket[c.charCodeAt(0) - 97].length;
    for (let i = 0; i < len; i++) {
      let [i, j] = bucket[c.charCodeAt(0) - 97].shift()!;
      if (j === words[i].length - 1) {
        res++;
      } else {
        j++;
        bucket[words[i].charCodeAt(j) - 97].push([i, j]);
      }
    }
  }
  return res;
};
// @lc code=end

File: leetcode/two-pointers/792.匹配子序列的单词数.rs
/*
 * @lc app=leetcode.cn id=792 lang=rust
 *
 * [792] 匹配子序列的单词数
 *
 * https://leetcode.cn/problems/number-of-matching-subsequences/description/
 *
 * algorithms
 * Medium (47.71%)
 * Likes:    260
 * Dislikes: 0
 * Total Accepted:    18.4K
 * Total Submissions: 37.4K
 * Testcase Example:  '"abcde"\n["a","bb","acd","ace"]'
 *
 * 给定字符串 s 和字符串数组 words, 返回  words[i] 中是s的子序列的单词个数 。
 *
 * 字符串的 子序列 是从原始字符串中生成的新字符串，可以从中删去一些字符(可以是none)，而不改变其余字符的相对顺序。
 *
 *
 * 例如， “ace” 是 “abcde” 的子序列。
 *
 *
 *
 *
 * 示例 1:
 *
 *
 * 输入: s = "abcde", words = ["a","bb","acd","ace"]
 * 输出: 3
 * 解释: 有三个是 s 的子序列的单词: "a", "acd", "ace"。
 *
 *
 * Example 2:
 *
 *
 * 输入: s = "dsahjpjauf", words = ["ahjpjau","ja","ahbwzgqnuk","tnmlanowax"]
 * 输出: 2
 *
 *
 *
 *
 * 提示:
 *
 *
 * 1 <= s.length <= 5 * 10^4
 * 1 <= words.length <= 5000
 * 1 <= words[i].length <= 50
 * words[i]和 s 都只由小写字母组成。
 *
 * ​​​​
 */

// @lc code=start
use std::collections::VecDeque;
impl Solution {
    pub fn num_matching_subseq(s: String, words: Vec<String>) -> i32 {
        let mut res = 0;
        let mut list = VecDeque::new();
        let mut bucket: Vec<VecDeque<(usize, usize)>> = vec![list; 26];
        for i in 0..words.len() {
            bucket[words[i].chars().nth(0).unwrap() as usize - 97].push_back((i, 0));
        }
        for ch in s.bytes() {
            let n = bucket[ch as usize - 97].len();
            for i in 0..n {
                if let Some((i, j)) = bucket[ch as usize - 97].pop_front() {
                    if j == words[i].len() - 1 {
                        res += 1;
                    } else {
                        let j = j + 1;
                        bucket[words[i].chars().nth(j).unwrap() as usize - 97].push_back((i, j))
                    }
                }
            }
        }
        res
    }
}
// @lc code=end
